### PR TITLE
EIP-4834 - Hierarchical Domains Standard

### PR DESCRIPTION
**Pull request by Pandapip1**
When opening a pull request to submit a new EIP, please use the suggested template: https://github.com/ethereum/EIPs/blob/master/eip-template.md

We have a GitHub bot that automatically merges some PRs. It will merge yours immediately if certain criteria are met:

 - The PR edits only existing draft PRs.
 - The build passes.
 - Your GitHub username or email address is listed in the 'author' header of all affected PRs, inside <triangular brackets>.
 - If matching on email address, the email address is the one publicly listed on your GitHub profile.

Link to @eth-bot comment (there were technical difficulties): https://github.com/ethereum/EIPs/pull/4835#issuecomment-1048458202

### CODE REVIEW

**Comment by Pandapip1** [comment]
Original draft: #4834

**Comment by Pandapip1** [comment]
Discussion link: https://ethereum-magicians.org/t/erc-4834-hierarchical-domains-standard/8388

**Comment by Pandapip1** [comment]
@eth-bot didn't add the newEIPFile label for some reason.

**Comment by Pandapip1** [comment]
@lightclient looks like you were meant to be pinged by the bot, but it didn't work.

**Comment by MicahZoltu** [comment]
@alita-moore Any idea why the bot didn't comment on this PR?

File: EIPS/eip-4834.md [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+Solidity Interface (with descriptions): [eip-4834/IDomain.sol]
+
+## Rationale
+As previously mentioned in the abstract and motivation, the design of this ERC is different than that of ERC-137. ERC-137 has a strong tiering - the registrars update the ENS registry, which points to resolvers. In ERC-4834, every domain is effectively a registry, resolver, and registrar all in one. This is specifically designed to be a universal addressing system.
+
+## Backwards Compatibility
+There is no compatibility with ERC-137. While the two ERCs are somewhat similar, the rationale for each is different. An ERC-4834 contract that wraps the ERC-137 contract could be created, however, that would support .eth domains, so compatibility is possible.
+
+## Reference Implementation
+Ownable (eip-173) domain: [eip-4834/OwnableDomain.sol]
```
Comment (MicahZoltu) [inline EIPS/eip-4834.md:34]:
```suggestion
Ownable (eip-173) domain: [./eip-4834/OwnableDomain.sol]
```

File: EIPS/eip-4834.md [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+
+This ERC also implements flexible access control. This allows for clever implementations, such as having a token that can be redeemed for subdomains of the token, and domains that are themselves NFTs that grant the ability to manage the domain to the owner of said NFT. There can also be multiple managers of a single domain with different access, instead of a single "owner."
+
+In addition, TLDs (like .com), subdomains, and even the "root" itself (analogous to the ENS registry) are all implemented as ERC-4834 smart contracts, meaning that name-resolution is a relatively simple recursive algorithm, and there is no limit to the amount of nesting that is possible (e.g. 0.1.2.3.4.5.6.7.8.9.a.b.c.d.e.f.g.h.i.j.k.l.m.n.o.p.q.r.s.t.u.v.w.x.y.z could be valid if the root contains z, and z contains y, and etc).
+
+## Specification
+Solidity Interface (with descriptions): [eip-4834/IDomain.sol]
```
Comment (MicahZoltu) [inline EIPS/eip-4834.md:25]: The specification should be inline.  You should be able to do the following in order to get syntax highlighting.
````
```sol
<specification here>
```
````

**Comment by alita-moore** [comment]
@MicahZoltu it's because it was draft and then went to review. So the bot hasn't actually run yet. Easiest fix is to add that event type to the file.

But this isn't critical because it can be fixed by closing the PR / reopening it or by new commits.

**Comment by alita-moore** [comment]
nvm, it should already be running when you open it for review: [ShixuanZhao](https://github.com/snackpass-interviews/Shixuan_Zhao/commits?author=ShixuanZhao)

**Comment by Pandapip1** [comment]
Maybe it just went down for a few hours - I saw another PR that failed to get tags.

File: EIPS/eip-4834.md (hunk 2 above)
Comment (Pandapip1) [reply EIPS/eip-4834.md]: Inlined it

**Comment by Pandapip1** [comment]
Also, I've modified the gitignore so that I can edit this in gitpod (https://gitpod.io). There's no way to turn off the creation of the .gitpod.yml file, so I've added it to the gitignore to avoid clutter. LMK if that's okay!

**Comment by Pandapip1** [comment]
Making a few changes, I'll put this in draft for now.

File: EIPS/eip-4834.md (hunk 2 above)
Comment (Pandapip1) [reply EIPS/eip-4834.md]: I'm going to unresolve this so I remember to put the inline interface back.

**Comment by Pandapip1** [comment]
Re-triggering @eth-bot

**Comment by Pandapip1** [comment]
Reviewer pings: @lightclient, @axic, @SamWilsn

(Let me know if it's okay to have added that file to the gitignore. Thanks!)

File: .gitignore [hunk 3]
```diff
@@ -2,3 +2,4 @@ _site
 .sass-cache
 .jekyll-metadata
 vendor
+.gitpod.yml
```
Comment (MicahZoltu) [inline .gitignore:5]: This `.gitignore` change should be a separate pull request that can be discussed in isolation from this EIP.  I recommend removing it from this EIP (you can just do it in the GitHub UI by accepting this suggestion), and then submitting a separate PR that *only* adds `.gitpod.yml` to the ignore list so we can discuss it properly, and the history of `.gitignore` isn't cluttered with EIPs.
```suggestion
```

Comment (Pandapip1) [reply .gitignore]: Created #4865

File: assets/eip-4834/OwnableDomain.sol [hunk 4]
```diff
@@ -0,0 +1,154 @@
...
+        require(this.hasDomain(name));
+        require(this.canDeleteDomain(msg.sender, name));
+
+        subdomainsPresent[name] = false; // Only need to mark it as deleted
+
+        emit SubdomainDelete(msg.sender, name, subdomains[name]);
+    }
+
+
+    //// Pare   nt Domain Access Control
```
Comment (Pandapip1) [inline assets/eip-4834/OwnableDomain.sol:86]: Typo

File: assets/eip-4834/NFTDomain.sol [hunk 5]
```diff
@@ -0,0 +1,159 @@
...
+        require(this.hasDomain(name));
+        require(this.canDeleteDomain(msg.sender, name));
+
+        subdomainsPresent[name] = false; // Only need to mark it as deleted
+
+        emit SubdomainDelete(msg.sender, name, subdomains[name]);
+    }
+
+
+    //// Pare   nt Domain Access Control
```
Comment (Pandapip1) [inline assets/eip-4834/NFTDomain.sol:87]: Typo

File: assets/eip-4834/NaiveDomain.sol [hunk 6]
```diff
@@ -0,0 +1,152 @@
...
+        require(this.hasDomain(name));
+        require(this.canDeleteDomain(msg.sender, name));
+
+        subdomainsPresent[name] = false; // Only need to mark it as deleted
+
+        emit SubdomainDelete(msg.sender, name, subdomains[name]);
+    }
+
+
+    //// Pare   nt Domain Access Control
```
Comment (Pandapip1) [inline assets/eip-4834/NaiveDomain.sol:84]: Typo

File: EIPS/eip-4834.md [hunk 7]
```diff
@@ -0,0 +1,184 @@
...
+discussions-to: https://ethereum-magicians.org/t/erc-4834-hierarchical-domains-standard/8388
+status: Draft
+type: Standards Track
+category: ERC
+created: 2022-02-22
+requires: 165
+---
+
+## Abstract
+The Hierarchical Domains Standard is a standard for extremely generic name resolution. conversely it makes it simple to extend - any smart contract that implements the ERC-4834 interface could be set as a subdomain. This differs from ERC-137 because it does not provide a separate "resolver" contract - any behavior should be part of the ERC-4834 contract itself.
```
Comment (Pandapip1) [inline EIPS/eip-4834.md:15]: Typo

**Comment by Pandapip1** [comment]
When is this likely to be reviewed?

**Comment by Pandapip1** [comment]
https://rinkeby.etherscan.io/address/0xecf183a137e28fc4b48a57d1559a7904f8df28b9

**Comment by MicahZoltu** [comment]
> When is this likely to be reviewed by the editors?

There are fewer editors who review ERCs (I'm not one of them), so it can sometimes take a week or two to get a review unfortunately.

**Comment by Pandapip1** [comment]
Travis is stuck, re-triggering

File: EIPS/eip-4834.md [hunk 8]
```diff
@@ -0,0 +1,224 @@
...
+Pop off the last element of the array (`"c"`), then call `domain.hasDomain(lastElement)`. If it's `false`, then the domain resolution fails. Otherwise, set the domain to `domain.getDomain(lastElement)`. Repeat until there are no more lists.
+
+Here is a solidity function that resolves a name:
+
+```sol
+function resolve(string[] calldata splitName, IDomain root) public view returns (address) {
+    IDomain current = root;
+    for (uint i = splitName.length - 1; i >= 0; i--) {
+        // Ensure that domain is actually a domain
+        require(current.supportsInterface(type(IDomain).interfaceId));
```
Comment (SamWilsn) [inline EIPS/eip-4834.md:165]: I think it's important to mention this check (`supportsInterface`) as part of the algorithm described above.

File: EIPS/eip-4834.md [hunk 9]
```diff
@@ -0,0 +1,224 @@
...
+
+### Supported Interfaces
+| Interface Name 	| ERC-165 Hash 	| Reference                                       	|
+|----------------	|--------------	|-------------------------------------------------	|
+
+## Backwards Compatibility
+There is no compatibility with ERC-137. While the two ERCs are somewhat similar, the rationale for each is different. It is also likely impossible to create an ERC-4834 domain that wraps ERC-137, because ERC-137 domains are indexed by the hash of the entire domain, while ERC-4834 domains are indexed by the subdomain name.
+
+## Reference Implementations
+Some of these implementations use some contracts from OpenZeppelin (included in the contracts directory) for ease of implementation and understandability.
```
Comment (SamWilsn) [inline EIPS/eip-4834.md:183]: > OpenZeppelin (included in the contracts directory)

where are these included?

File: EIPS/eip-4834.md [hunk 10]
```diff
@@ -0,0 +1,224 @@
...
+The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD", "SHOULD NOT", "RECOMMENDED", "MAY", and "OPTIONAL" in this document are to be interpreted as described in RFC 2119.
+
+### Contract Interface
+Solidity Interface with NatSpec & OpenZeppelin v4 Interfaces (also available at [IDomain.sol](../assets/eip-4834/IDomain.sol)):
+
+```sol
+interface IDomain is IERC165 {
+    //// Events
+    
+    /// @notice     Must be emitted when createDomain is called
```
Comment (SamWilsn) [inline EIPS/eip-4834.md:38]:
```suggestion
    /// @notice     Must be emitted when a new subdomain is created (eg. through `createDomain`.)
```

I worry that the original wording is too specific, and would let implementations weasel out of emitting the events.

File: EIPS/eip-4834.md [hunk 11]
```diff
@@ -0,0 +1,224 @@
...
+    /// @param      sender msg.sender for createDomain
+    /// @param      name name for createDomain
+    /// @param      subdomain subdomain in createDomain
+    event SubdomainCreate(address indexed sender, string name, address subdomain);
+
+    /// @notice     Must be emitted when setDomain is called
```
Comment (SamWilsn) [inline EIPS/eip-4834.md:44]:
```suggestion
    /// @notice     Must be emitted when the resolved address for a domain is changed (eg. with `setDomain`)
```

File: EIPS/eip-4834.md [hunk 12]
```diff
@@ -0,0 +1,224 @@
...
+    /// @param      sender msg.sender for setDomain
+    /// @param      name name for setDomain
+    /// @param      subdomain subdomain in setDomain
+    /// @param      oldSubdomain the old subdomain
+    event SubdomainUpdate(address indexed sender, string name, address subdomain, address oldSubdomain);
+
+    /// @notice     Must be emitted when deleteDomain is called
```
Comment (SamWilsn) [inline EIPS/eip-4834.md:51]:
```suggestion
    /// @notice     Must be emitted when a domain is unmapped (eg. with `deleteDomain`)
```

File: EIPS/eip-4834.md [hunk 13]
```diff
@@ -0,0 +1,224 @@
...
+    ///             `getDomain(name).canDeleteSubdomain(msg.sender, this, subdomain)` is `false`.
+    /// @param      updater The account that may or may not be able to delete a subdomain
+    /// @param      name The subdomain to delete
+    /// @return     Whether an account can delete the subdomain
+    function canDeleteDomain(address updater, string memory name) external view returns (bool);
+
+
+    //// Subdomain Access Control
+
+    /// @notice     Get if an account can point this domain as a subdomain
```
Comment (SamWilsn) [inline EIPS/eip-4834.md:126]: I'm not sure what this means. Would you mind expanding it a little bit?

File: EIPS/eip-4834.md [hunk 14]
```diff
@@ -0,0 +1,224 @@
...
+        // Resolve subdomain
+        current = current.getDomain(splitName[i]);
+    }
+    return current;
+}
+```
+
+### Supported Interfaces
+| Interface Name 	| ERC-165 Hash 	| Reference                                       	|
+|----------------	|--------------	|-------------------------------------------------	|
```
[the remark this answers is not in the document]
Comment (Pandapip1) [reply EIPS/eip-4834.md:177]: No - see EIP137

File: EIPS/eip-4834.md (hunk 9 above)
Comment (Pandapip1) [reply EIPS/eip-4834.md:183]: I forgot to remove that, thanks!

File: EIPS/eip-4834.md (hunk 8 above)
Comment (Pandapip1) [reply EIPS/eip-4834.md:165]: The check was just because I wasn't sure what would happen. I'll remove it since I'm pretty sure it won't break.

File: EIPS/eip-4834.md [hunk 15]
```diff
@@ -0,0 +1,224 @@
...
+
```
[the remark this answers is not in the document]
Comment (Pandapip1) [reply EIPS/eip-4834.md:178]: Huh, I thought I added one.

[the remark this answers is not in the document]
Comment (Pandapip1) [reply EIPS/eip-4834.md:178]: Oh, I accidentally forgot to re-paste it in last commit.

File: EIPS/eip-4834.md [hunk 16]
```diff
@@ -0,0 +1,224 @@
...
+    /// @notice     Get if an account can move the subdomain away from the current domain
+    /// @dev        May be called by `canSetDomain` of the parent domain - implement access control here!!!
+    /// @param      updater The account that may be moving the subdomain
+    /// @param      name The subdomain name
+    /// @param      parent The parent domain
+    /// @param      newSubdomain The domain that will be set next
+    /// @return     Whether an account can update the subdomain
+    function canMoveSubdomain(address updater, string memory name, IDomain parent, address newSubdomain) external view returns (bool);
+
+    /// @notice     Get if an account can point this domain as a subdomain
```
[the remark this answers is not in the document]
Comment (Pandapip1) [reply EIPS/eip-4834.md]: Yup, I must have missed that.

File: EIPS/eip-4834.md (hunk 13 above)
Comment (Pandapip1) [reply EIPS/eip-4834.md]: I've rephrased it: `Get if an account can set this domain as a subdomain`

**Comment by Pandapip1** [comment]
Okay, the impls *should* work now.

**Comment by Pandapip1** [comment]
So the CI's angry at OpenZeppelin - any suggestions?

**Comment by MicahZoltu** [comment]
It does appear that way.  That being said, OZ contracts aren't CC0 so you can't include them anyway.  Also, a reference implementation should, ideally, be very small and not necessarily production complete.  This means you don't need the vast majority of features (security, modularity, etc.) that OZ contracts provide.

I recommend trimming down the reference implementation greatly such that it *only* exemplifies the things that matter for this EIP.  It doesn't need to be complete or even fully functional, you can stub out things that are out of scope for this EIP.

Separately: Someone should probably tell OZ about all of their spelling errors.  😆 

```
- ./_site/assets/eip-4834/OpenZeppelin/DOCUMENTATION.html
  *  internally linking to /docs, which does not exist (line 61)
     <a href="/docs">docs</a>
```
```
./assets/eip-4834/OpenZeppelin/package-lock.json:9342: vEw ==> view
./assets/eip-4834/OpenZeppelin/package-lock.json:16026: vEw ==> view
./assets/eip-4834/OpenZeppelin/package-lock.json:16641: vEw ==> view
./assets/eip-4834/OpenZeppelin/package-lock.json:26020: vEw ==> view
./assets/eip-4834/OpenZeppelin/package-lock.json:31325: vEw ==> view
./assets/eip-4834/OpenZeppelin/package-lock.json:31825: vEw ==> view
./assets/eip-4834/OpenZeppelin/CHANGELOG.md:459: libray ==> library
./assets/eip-4834/OpenZeppelin/certora/applyHarness.patch:74: consistant ==> consistent
./assets/eip-4834/OpenZeppelin/certora/specs/GovernorBase.spec:176: completness ==> completeness
./assets/eip-4834/OpenZeppelin/certora/specs/GovernorBase.spec:177: chages ==> changes, charges
./assets/eip-4834/OpenZeppelin/certora/specs/GovernorBase.spec:180: seperately ==> separately
./assets/eip-4834/OpenZeppelin/certora/specs/GovernorCountingSimple.spec:131: completness ==> completeness
./assets/eip-4834/OpenZeppelin/certora/specs/GovernorCountingSimple.spec:132: chages ==> changes, charges
./assets/eip-4834/OpenZeppelin/certora/specs/GovernorCountingSimple.spec:135: seperately ==> separately
./assets/eip-4834/OpenZeppelin/certora/specs/GovernorCountingSimple.spec:208: priveleged ==> privileged
./assets/eip-4834/OpenZeppelin/certora/specs/GovernorCountingSimple.spec:220: priveleged ==> privileged
./assets/eip-4834/OpenZeppelin/contracts/interfaces/IERC2981.sol:19: payed ==> paid
./assets/eip-4834/OpenZeppelin/contracts/utils/structs/DoubleEndedQueue.sol:27: could't ==> couldn't
./assets/eip-4834/OpenZeppelin/contracts/utils/math/SafeMath.sol:31: substraction ==> subtraction
./assets/eip-4834/OpenZeppelin/contracts/proxy/Proxy.sol:48: overriden ==> overridden
./assets/eip-4834/OpenZeppelin/contracts/proxy/Proxy.sol:83: overriden ==> overridden
./assets/eip-4834/OpenZeppelin/contracts/governance/README.adoc:43: requiering ==> requiring
./assets/eip-4834/OpenZeppelin/contracts/governance/TimelockController.sol:188: datas ==> data
./assets/eip-4834/OpenZeppelin/contracts/governance/TimelockController.sol:192: datas ==> data
./assets/eip-4834/OpenZeppelin/contracts/governance/TimelockController.sol:229: datas ==> data
./assets/eip-4834/OpenZeppelin/contracts/governance/TimelockController.sol:235: datas ==> data
./assets/eip-4834/OpenZeppelin/contracts/governance/TimelockController.sol:237: datas ==> data
./assets/eip-4834/OpenZeppelin/contracts/governance/TimelockController.sol:240: datas ==> data
./assets/eip-4834/OpenZeppelin/contracts/governance/TimelockController.sol:304: datas ==> data
./assets/eip-4834/OpenZeppelin/contracts/governance/TimelockController.sol:309: datas ==> data
./assets/eip-4834/OpenZeppelin/contracts/governance/TimelockController.sol:311: datas ==> data
./assets/eip-4834/OpenZeppelin/contracts/governance/TimelockController.sol:314: datas ==> data
./assets/eip-4834/OpenZeppelin/contracts/governance/Governor.sol:49: consummed ==> consumed
./assets/eip-4834/OpenZeppelin/contracts/governance/Governor.sol:67: poping ==> popping, pooping
./assets/eip-4834/OpenZeppelin/contracts/governance/Governor.sol:127: accross ==> across
./assets/eip-4834/OpenZeppelin/contracts/governance/Governor.sol:232: overriden ==> overridden
./assets/eip-4834/OpenZeppelin/contracts/governance/Governor.sol:311: overriden ==> overridden
./assets/eip-4834/OpenZeppelin/contracts/governance/Governor.sol:328: trigerred ==> triggered
./assets/eip-4834/OpenZeppelin/contracts/governance/Governor.sol:347: trigerred ==> triggered
./assets/eip-4834/OpenZeppelin/contracts/governance/IGovernor.sol:161: snaphot ==> snapshot
./assets/eip-4834/OpenZeppelin/contracts/governance/IGovernor.sol:162: quroum ==> quorum
./assets/eip-4834/OpenZeppelin/contracts/governance/extensions/GovernorTimelockControl.sol:50: Overriden ==> Overridden
./assets/eip-4834/OpenZeppelin/contracts/governance/extensions/GovernorTimelockControl.sol:110: Overriden ==> Overridden
./assets/eip-4834/OpenZeppelin/contracts/governance/extensions/GovernorTimelockControl.sol:123: Overriden ==> Overridden
./assets/eip-4834/OpenZeppelin/contracts/governance/extensions/GovernorTimelockCompound.sol:108: Overriden ==> Overridden
./assets/eip-4834/OpenZeppelin/contracts/governance/extensions/GovernorTimelockCompound.sol:170: Overriden ==> Overridden
./assets/eip-4834/OpenZeppelin/contracts/governance/extensions/GovernorTimelockCompound.sol:188: Overriden ==> Overridden
./assets/eip-4834/OpenZeppelin/contracts/finance/VestingWallet.sol:123: amout ==> amount
./assets/eip-4834/OpenZeppelin/test/security/Pausable.test.js:70: process' ==> process's
./assets/eip-4834/OpenZeppelin/test/helpers/governance.js:109: otherwize ==> otherwise
./assets/eip-4834/OpenZeppelin/test/helpers/governance.js:115: otherwize ==> otherwise
./assets/eip-4834/OpenZeppelin/test/helpers/governance.js:168: queing ==> queueing
./assets/eip-4834/OpenZeppelin/test/utils/structs/DoubleEndedQueue.test.js:72: begining ==> beginning
./assets/eip-4834/OpenZeppelin/test/utils/cryptography/ECDSA.test.js:50: siganture ==> signature
./assets/eip-4834/OpenZeppelin/test/governance/TimelockController.test.js:28: datas ==> data
./assets/eip-4834/OpenZeppelin/test/governance/TimelockController.test.js:38: datas ==> data
./assets/eip-4834/OpenZeppelin/test/governance/TimelockController.test.js:42: datas ==> data
./assets/eip-4834/OpenZeppelin/test/governance/TimelockController.test.js:122: datas ==> data
./assets/eip-4834/OpenZeppelin/test/governance/TimelockController.test.js:166: overwritting ==> overwriting
./assets/eip-4834/OpenZeppelin/test/governance/TimelockController.test.js:349: datas ==> data
./assets/eip-4834/OpenZeppelin/test/governance/TimelockController.test.js:361: datas ==> data
./assets/eip-4834/OpenZeppelin/test/governance/TimelockController.test.js:373: overwritting ==> overwriting
./assets/eip-4834/OpenZeppelin/test/governance/TimelockController.test.js:377: datas ==> data
./assets/eip-4834/OpenZeppelin/test/governance/TimelockController.test.js:388: datas ==> data
./assets/eip-4834/OpenZeppelin/test/governance/TimelockController.test.js:403: datas ==> data
./assets/eip-4834/OpenZeppelin/test/governance/TimelockController.test.js:433: datas ==> data
./assets/eip-4834/OpenZeppelin/test/governance/TimelockController.test.js:448: datas ==> data
./assets/eip-4834/OpenZeppelin/test/governance/TimelockController.test.js:475: datas ==> data
./assets/eip-4834/OpenZeppelin/test/governance/TimelockController.test.js:489: datas ==> data
./assets/eip-4834/OpenZeppelin/test/governance/TimelockController.test.js:502: datas ==> data
./assets/eip-4834/OpenZeppelin/test/governance/TimelockController.test.js:519: datas ==> data
./assets/eip-4834/OpenZeppelin/test/governance/TimelockController.test.js:538: datas ==> data
./assets/eip-4834/OpenZeppelin/test/governance/TimelockController.test.js:549: datas ==> data
./assets/eip-4834/OpenZeppelin/test/governance/TimelockController.test.js:559: datas ==> data
./assets/eip-4834/OpenZeppelin/test/governance/TimelockController.test.js:573: datas ==> data
./assets/eip-4834/OpenZeppelin/test/governance/TimelockController.test.js:587: datas ==> data
./assets/eip-4834/OpenZeppelin/test/governance/TimelockController.test.js:636: datas ==> data
./assets/eip-4834/OpenZeppelin/test/governance/TimelockController.test.js:647: datas ==> data
./assets/eip-4834/OpenZeppelin/test/governance/compatibility/GovernorCompatibilityBravo.test.js:218: doesnt ==> doesn't, does not
./assets/eip-4834/OpenZeppelin/audit/2017-03.md:136: modifers ==> modifiers
```

**Comment by Pandapip1** [comment]
> That being said, OZ contracts aren't CC0 so you can't include them anyway. 

I don't see the issue. They're MIT licence, so you're fully allowed to include them if you keep the LICENCE file (as per the LICENCE file itself). It's also clear, from them being a submodule, that they are a different repository and therefore are possibly files with a different licence.

I'd *like* a backup of OpenZeppelin in there, just because of no external links. Let me know if you agree.

If not, I might open a PR adding a reference implementation to EIP 721, another PR for EIP 165, and reference those in my implementations.

Also: https://github.com/OpenZeppelin/openzeppelin-contracts/issues/3270 🙂

**Comment by MicahZoltu** [comment]
Ah, I didn't realize you added it as a submodule.  I definitely don't think we should be adding submodules.  That suffers basically the exact same problem as external links has, and it adds a *lot* of complexity to the repository.

I still think this reference implementation is *too* complete.  A good reference implementation should be super easy to read, and that (IMO) should trump things like security, safety, etc.  People should not be using reference implementations in production, they should just be using them to better understand how something works.  As an example, one could just have `contract Apple is IERC20` and then simply not define `IERC20` anywhere.  This keeps this repository free from encumbrance with writing a token contract so it can focus exclusively on what this EIP brings to the table.

**Comment by Pandapip1** [comment]
> I definitely don't think we should be adding submodules.

Can I open a PR to add this to EIP-1? I would have thought that submodules would be a good way of doing this, as it copies the best of external links with and copies of the actual repository.

**Comment by MicahZoltu** [comment]
Submodules don't copy the repository, they just link to it and like any other link that can 404 with time.

**Comment by Pandapip1** [comment]
> Submodules don't copy the repository, they just link to it and like any other link that can 404 with time.

IIUC that isn't the case. The git repo is still there, but github displays it as a link.

Anyways, I agree with dropping OpenZeppelin - although it's really nice it's not irreplaceable.

**Comment by MicahZoltu** [comment]
> IIUC that isn't the case. The git repo is still there, but github displays it as a link.

When you do a git clone of a project with submodules, you won't get the submodules by default.  You have to follow it up with other commands to clone the submodule (from its original git repository.  Check out the section here (https://git-scm.com/book/en/v2/Git-Tools-Submodules#_cloning_submodules) in particular notice how after `git clone` the subdirectory where the submodule would be is empty at first (0 bytes on disk).  It isn't until you run `git submodule init` and `git submodule update` that you actually fetch the contents, and they are fetched from wherever the submodule is pointing at (in this case, `github.com/OpenZeppelin/contracts`).

File: EIPS/eip-4834.md [hunk 17]
```diff
@@ -0,0 +1,221 @@
...
+
+In addition, TLDs (like .com), subdomains, and even the "root" itself (analogous to the ENS registry) are all implemented as ERC-4834 smart contracts, meaning that name-resolution is a relatively simple iterative algorithm, and there is no limit to the amount of nesting that is possible (e.g. 0.1.2.3.4.5.6.7.8.9.a.b.c.d.e.f.g.h.i.j.k.l.m.n.o.p.q.r.s.t.u.v.w.x.y.z could be valid if the root contains z, and z contains y, and etc).
+
+## Rationale
+As previously mentioned in the abstract and motivation, the design of this ERC is different than that of ERC-137. ERC-137 has a strong tiering - the registrars update the ENS registry, which points to resolvers. In ERC-4834, every domain is effectively a registry, resolver, and registrar all in one. This is specifically designed to be a simple addressing system.
```
Comment (SamWilsn) [inline EIPS/eip-4834.md:29]: `Rationale` has to go after `Specification`.

Also, the `Rationale` section should describe the reasons behind particular choices in the specification (a dumb example would be to explain why `canDeleteDomain` and `canDeleteSubdomain` both exist.) Compare that to the `Motivation` section, which describes why this EIP as a whole should exist.

File: EIPS/eip-4834.md [hunk 18]
```diff
@@ -0,0 +1,221 @@
...
+        current = current.getDomain(splitName[i]);
+    }
+    return current;
+}
+```
+
+<!--### Supported Interfaces
+| Interface Name 	| ERC-165 Hash 	| Reference                                       	|
+|----------------	|--------------	|-------------------------------------------------	|-->
+
```
Comment (SamWilsn) [inline EIPS/eip-4834.md:175]: Should probably remove this before merging.

File: EIPS/eip-4834.md [hunk 19]
```diff
@@ -0,0 +1,221 @@
...
+discussions-to: https://ethereum-magicians.org/t/erc-4834-hierarchical-domains-standard/8388
+status: Draft
+type: Standards Track
+category: ERC
+created: 2022-02-22
+requires: 165
+---
+
+## Abstract
+`ERC-4834` is a standard for generic name resolution with access control. It permits contracts that implement `ERC-4834` ("domains") to be addressable with a more human-friendly name, similar to `ERC-137`.
```
Comment (SamWilsn) [inline EIPS/eip-4834.md:15]: Oh, sorry, I forgot to mention that you have to link to an EIP the first time you mention it.

File: EIPS/eip-4834.md (hunk 18 above)
Comment (Pandapip1) [reply EIPS/eip-4834.md:175]: I have it commented out -- is that good enough?

File: EIPS/eip-4834.md (hunk 19 above)
Comment (Pandapip1) [reply EIPS/eip-4834.md]: Done

**Comment by Pandapip1** [comment]
@SamWilsn mind re-reviewing? (The re-request review button still isn't working)

**Comment by Pandapip1** [comment]
Bump.

File: EIPS/eip-4834.md [hunk 20]
```diff
@@ -0,0 +1,230 @@
+---
+eip: 4834
+title: Hierarchical Domain Standard
+description: The Hierarchal Domain Standard is a standard for extremely generic name resolution.
```
Comment (lightclient) [inline EIPS/eip-4834.md:4]:
```suggestion
description: A standard for extremely generic name resolution.
```

File: EIPS/eip-4834.md [hunk 21]
```diff
@@ -0,0 +1,230 @@
...
+discussions-to: https://ethereum-magicians.org/t/erc-4834-hierarchical-domains-standard/8388
+status: Draft
+type: Standards Track
+category: ERC
+created: 2022-02-22
+requires: 20, 137, 165, 1155
+---
+
+## Abstract
+This EIP, `EIP-4834` is a standard for generic name resolution with access control. It permits a contract that implements `EIP-4834` (referred to as a "domain" hereafter) to be addressable with a more human-friendly name, with a similar purpose to [`EIP-137`](./eip-137) (referred to as "ENS" hereafter).
```
Comment (lightclient) [inline EIPS/eip-4834.md:15]:
```suggestion
This is a standard for generic name resolution with access control. It permits a contract that implements `EIP-4834` (referred to as a "domain" hereafter) to be addressable with a more human-friendly name, with a similar purpose to [`EIP-137`](./eip-137) (referred to as "ENS" hereafter).
```

**Review by lightclient — CHANGES_REQUESTED** [review]

**Review by lightclient — APPROVED** [review]
Overall looks good. I do see that you're requiring EIP-1155, but I don't see any occurrence of it in your spec?
